beacon-chain/db/slasherkv/slasher.go

```diff
@@ -278,60 +289,117 @@ func (s *Store) SaveAttestationRecordsForValidators(
+		start := max(0, stop-batchSize)
 
-		for i := len(attestations) - 1; i >= 0; i-- {
-			attestation := attestations[i]
+		attWrappersBatch := attWrappers[start:stop]
+		encodedTargetEpochBatch := encodedTargetEpoch[start:stop]
+		encodedRecordsBatch := encodedRecords[start:stop]
 
-			if err := attRecordsBkt.Put(attestation.DataRoot[:], encodedRecords[i]); err != nil {
-				return err
-			}
+		// Perform basic check.
+		if len(encodedTargetEpochBatch) != len(encodedRecordsBatch) {
+			return fmt.Errorf(
+				"cannot save attestation records, got %d target epochs and %d records",
+				len(encodedTargetEpochBatch), len(encodedRecordsBatch),
+			)
+		}
+
+		currentBatchSize := len(encodedTargetEpochBatch)
+
+		// Save attestation records in the database.
+		if err := s.db.Update(func(tx *bolt.Tx) error {
+			attRecordsBkt := tx.Bucket(attestationRecordsBucket)
+			dataRootsBkt := tx.Bucket(attestationDataRootsBucket)
+
+			for i := currentBatchSize - 1; i >= 0; i-- {
+				attWrapper := attWrappersBatch[i]
+				dataRoot := attWrapper.DataRoot
 
-			for _, valIdx := range attestation.IndexedAttestation.AttestingIndices {
-				encIdx := encodeValidatorIndex(primitives.ValidatorIndex(valIdx))
+				encodedTargetEpoch := encodedTargetEpochBatch[i]
+				encodedRecord := encodedRecordsBatch[i]
 
-				key := append(encodedTargetEpoch[i], encIdx...)
-				if err := dataRootsBkt.Put(key, attestation.DataRoot[:]); err != nil {
+				if err := attRecordsBkt.Put(dataRoot[:], encodedRecord); err != nil {
 					return err
 				}
+
+				for _, validatorIndex := range attWrapper.IndexedAttestation.AttestingIndices {
+					encodedIndex := encodeValidatorIndex(primitives.ValidatorIndex(validatorIndex))
+
+					key := append(encodedTargetEpoch, encodedIndex...)
+					if err := dataRootsBkt.Put(key, dataRoot[:]); err != nil {
+						return err
+					}
+				}
 			}
+
+			return nil
+		}); err != nil {
+			return errors.Wrap(err, "failed to save attestation records")
 		}
+	}
 
-		return nil
-	})
+	return nil
 }
 
 // LoadSlasherChunks given a chunk kind and a disk keys, retrieves chunks for a validator
 // min or max span used by slasher from our database.
 func (s *Store) LoadSlasherChunks(
-	ctx context.Context, kind slashertypes.ChunkKind, diskKeys [][]byte,
+	ctx context.Context, kind slashertypes.ChunkKind, chunkKeys [][]byte,
 ) ([][]uint16, []bool, error) {
 	_, span := trace.StartSpan(ctx, "BeaconDB.LoadSlasherChunk")
 	defer span.End()
-	chunks := make([][]uint16, 0)
-	var exists []bool
-	err := s.db.View(func(tx *bolt.Tx) error {
-		bkt := tx.Bucket(slasherChunksBucket)
-		for _, diskKey := range diskKeys {
-			key := append(ssz.MarshalUint8(make([]byte, 0), uint8(kind)), diskKey...)
-			chunkBytes := bkt.Get(key)
-			if chunkBytes == nil {
-				chunks = append(chunks, []uint16{})
-				exists = append(exists, false)
-				continue
-			}
-			chunk, err := decodeSlasherChunk(chunkBytes)
-			if err != nil {
-				return err
+
+	const chunkSize = 10_000
+
```

Comment:
same here

Reply:
Fixed in a8dcbe240e16b34c6740b5b2a1705b71c0376239.